libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -571,32 +585,48 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			else
 				args.emplace_back(convert(*arguments[i], *parameterTypes[i]).commaSeparatedList());
 
-		if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+		FunctionDefinition const* functionDef;
+		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+		{
+			solUnimplementedAssert(!functionType->bound(), "Internal calls to bound library functions are not yet implemented");
+
+			functionDef = dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration);
+			solAssert(functionDef, "Libraries cannot have state variables");
+			solAssert(!functionDef->virtualSemantics(), "Libraries can neither inherit nor be inherited");
+		}
+		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
 		{
 			solAssert(!functionType->bound(), "");
-			if (auto functionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
-			{
-				define(_functionCall) <<
-					m_context.enqueueFunctionForCodeGeneration(
-						functionDef->resolveVirtual(m_context.mostDerivedContract())
-					) <<
-					"(" <<
-					joinHumanReadable(args) <<
-					")\n";
-				return;
-			}
+
+			if (auto unresolvedFunctionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
+				functionDef = &unresolvedFunctionDef->resolveVirtual(m_context.mostDerivedContract());
+			else
+				// Not a function name - must be the name of a function pointer variable
+				functionDef = nullptr;
 		}
+		else
+			// Not a simple expression like x or A.x - we must be evaluating a function pointer
+			functionDef = nullptr;
 
-		define(_functionCall) <<
-			// NOTE: internalDispatch() takes care of adding the function to function generation queue
-			m_context.internalDispatch(
-				TupleType(functionType->parameterTypes()).sizeOnStack(),
-				TupleType(functionType->returnParameterTypes()).sizeOnStack()
-			) <<
-			"(" <<
-			IRVariable(_functionCall.expression()).part("functionIdentifier").name() <<
-			joinHumanReadablePrefixed(args) <<
-			")\n";
+		if (functionDef)
+			define(_functionCall) <<
+				m_context.enqueueFunctionForCodeGeneration(*functionDef) <<
+				"(" <<
+				joinHumanReadable(args) <<
+				")\n";
+		else
+			// FIXME: Pointers to internal library functions are supported in Solidity but internalDispatch()
```

Comment:
Do you think that we will look at this code line when we notice it again and implement it?
Maybe it's better to remove this or add a "fixme" into internalDispatch instead.

Reply:
I will remember because I'm going to implement it next :)

Honestly, I really wanted to use `solUnimplementedAssert()` here but could not find a way to discern calls to library functions from calls to contract functions when the call goes via a pointer.

Also I mostly wanted it to be noticed by reviewers, in the context of the whole PR, because in my opinion it's not obvious that this does not work yet.

Reply:
ok!

Reply:
If you just want to tell something to the reviewers, you can also use comments on the PR.

Reply:
I mean, I did :) It was just one of the extra reasons.

Ok, I agree that this is not the best place for this comment. I moved it to `internalDispatch()`.